app/src/main/java/pm/gnosis/heimdall/ui/safe/pairing/PairingAuthenticator.kt

```diff
@@ -0,0 +1,172 @@
+import pm.gnosis.heimdall.data.repositories.TokenRepository
+import pm.gnosis.heimdall.data.repositories.TransactionExecutionRepository
+import pm.gnosis.heimdall.data.repositories.models.ERC20Token
+import pm.gnosis.heimdall.data.repositories.models.ERC20TokenWithBalance
+import pm.gnosis.heimdall.data.repositories.models.SafeTransaction
+import pm.gnosis.heimdall.di.modules.ApplicationModule
+import pm.gnosis.heimdall.ui.safe.helpers.RecoverSafeOwnersHelper
+import pm.gnosis.model.Solidity
+import timber.log.Timber
+import java.math.BigInteger
+import java.math.RoundingMode
+import javax.inject.Inject
+
+abstract class PairingAuthenticatorContract : ViewModel() {
+
+    abstract val observableState: LiveData<ViewUpdate>
+
+    abstract fun setup(safeAddress: Solidity.Address)
+
+    abstract fun estimate()
+
+    /**
+     * @param signingSafe to indicate which key should be used for pairing. @null if a new key should be generated
+     */
+    abstract fun pair(payload: String)
+
+    sealed class PairingResult {
+
+        data class PairingError(val error: Exception) : PairingResult()
+
+        data class PairingSuccess(val safe: Solidity.Address, val extension: Solidity.Address) : PairingResult()
+    }
+
+    sealed class ViewUpdate(open val isLoading: Boolean) {
+
+        data class Balance(
+            override val isLoading: Boolean,
+            val balanceBefore: String,
+            val fee: String,
+            val balanceAfter: String,
+            val tokenSymbol: String,
+            val sufficient: Boolean
+        ) : ViewUpdate(isLoading)
+
+        data class Pairing(
+            override val isLoading: Boolean,
+            val pairingResult: PairingResult? = null
+        ) : ViewUpdate(isLoading)
+    }
+
+}
+
+class PairingAuthenticatorViewModel @Inject constructor(
+    private val recoverSafeOwnersHelper: RecoverSafeOwnersHelper,
+    private val gnosisSafeRepository: GnosisSafeRepository,
+    private val tokenRepository: TokenRepository,
+    private val transactionExecutionRepository: TransactionExecutionRepository,
+    private val pushServiceRepository: PushServiceRepository,
+    private val moshi: Moshi,
+    private val appDispatcher: ApplicationModule.AppCoroutineDispatchers
+) : PairingAuthenticatorContract() {
+
+    private val errorHandler = CoroutineExceptionHandler { _, e ->
+        viewModelScope.launch {
+            Timber.e(e)
+        }
+    }
+
+    override val observableState: LiveData<ViewUpdate>
+        get() = _state
+    private val _state = MutableLiveData<ViewUpdate>()
+
+    private lateinit var safeAddress: Solidity.Address
+
+    private lateinit var paymentToken: ERC20Token
+    private lateinit var pairingTransaction: SafeTransaction
+
+
+    override fun setup(safeAddress: Solidity.Address) {
+        this.safeAddress = safeAddress
+        runBlocking(appDispatcher.background + errorHandler) {
+            loadPaymentToken()
+            buildPairingTransaction()
```

Comment:
Why do we have to do this here vs doing it in the estimate function

Reply:
Same for the payment token

Reply:
we have to do it only once. Payment token and pairing transaction remain same. Only gas price estimate can change

Reply:
`buildPairingTransaction` is performing a network call that might fail and there is no way to retry this right now. Also payment token might actualy change, since you can change the payment token in the screen and currently it would not be reloaded.

Reply:
moved payment token & pairing transaction calls to estimate method